### DISCUSSION TITLE
Better automatic package creation and update

### DISCUSSION THREAD

**Discussion by adamjstewart** [spack/spack#31413] [Ideas]
### Summary

Allow me to summarize the feature I would like to see by walking you through how it would work. First, a user runs a command like:
```console
$ spack create -n foo -t python
```
Here we tell Spack that we want to create a package for `foo` and that it's a Python library. We haven't specified a URL, but Spack is smart enough to know that it should check PyPI to find the package. Spack looks on PyPI and finds 50 versions of foo available. It asks the user how many they would like to download. If the user chooses 5 versions, Spack downloads all 5 versions and computes the checksums.

Next, Spack checks the downloads for all 5 versions and automatically extracts the dependencies and version range for each version of foo. It also notices that there are some optional variants available and adds those variants and the dependencies they add.

A few months later, a new release comes out. Spackbot immediately notices this and runs the following command:
```console
$ spack checksum py-foo 1.2.3
```
This command downloads the new version, adds the checksum to the package, and updates any dependencies as needed. Spackbot then opens a PR to update this recipe. Maintainers can check it and see if anything is missing, otherwise it is quick and easy to approve and merge.

### Rationale

Creating and updating packages is very tedious. The author (and reviewer) need to manually inspect the source code, scour multiple files to determine the dependencies, and check to see if any of those dependencies have been updated. They also need to determine if the build system has any variants, what possible values those variants have, and write installation phases that implement these variants. We should automate this.

This has already been attempted for multiple build systems:
* **R**: @glennpj has scripts to do this for all R packages in Spack (both CRAN and BioConductor)
* **CMake**: Preliminary work done by @DavidPoliakoff in #3626 (no I haven't forgotten about this PR 😄).
* **Python**: Should be possible if we use wheels (https://github.com/spack/spack/discussions/30795).

@glennpj can you share your scripts here? Basically, I want to implement these directly in Spack for multiple build systems.

Overall, this dream will be a ton of work to get working properly, but @glennpj and I already spend countless hours maintaining the R and Python ecosystems in Spack. If this stuff was automated, even if it took us months to get it working, we could save a lot of time in the long run.

### Implementation

Of course, the devil is in the details. There are several things that this dream requires:

- [ ] Automatically determine where to find a package without a URL given its name and build system
- [ ] Extract tarballs (at least a single file) for each release
- [ ] Detect dependencies and variants (build system specific)
- [ ] Write these to packages using jinja2 templates
- [ ] Allow `spack checksum` to update packages, not just print what needs to be added
- [ ] Recursively add packages for dependencies that do not yet exist?
- [ ] Teach spackbot to detect new versions of packages and open PRs (dependabot and conda-forge-linter already do this)

This system will likely work for the vast majority of R, Python, Perl, and Octave packages since there is a uniform system for listing dependencies.

The templates in `spack create` are starting to get very complex. @alalazo uses jinja2 templates for module files. We should try to convert `spack create` to use jinja2 as well. I have some basic knowledge of how to use jinja2 but need to see how the module stuff works in Spack.

One thing we need to be careful of is that some packages require manual tweaks to add missing dependencies (for example, non-Python dependencies). We don't want to remove these lines, we just want to update them. @glennpj how do you handle this with R?

@tgamblin will love this

**Comment by alalazo** [comment]
I like the idea. I think refactor the `spack create` command to use Jinja 2 templates is already doable, so we can start from there if it makes sense to you. Also, since this is a very nice high-level description we might turn it into a [tasklist](https://docs.github.com/en/issues/tracking-your-work-with-issues/about-task-lists) kind of Issue and have a project which is specific to this feature and might span multiple releases.

**Reply by adamjstewart** [reply]
Yes, I think the jinja2 refactor would probably be the best place to start. I'm not sure if it's possible to use that in `spack checksum` or if it would only work for `spack create`. But we can start by using it in `spack create` and then it should be easier to auto-create packages. Auto-updating may be more tricky.

**Comment by glennpj** [comment]
This is something that I have been thinking about, particularly when doing a round of R package updates. I do have scripts for updating CRAN and BioConductor packages for Spack. I plan on contributing those when I have them in a form that would be usable by someone other than me. I still have to have eyeballs on the changing dependencies and constraints so they are not near the automation level that I would like. The recent addition of `with when()` syntax I think will help but I need to play with that more. In any event, being able to use the scripts for other build systems would be a stretch as they are very specific to R packages.

> One thing we need to be careful of is that some packages require manual tweaks to add missing dependencies (for example, non-Python dependencies). We don't want to remove these lines, we just want to update them. @glennpj how do you handle this with R?

Like I said, I still have to have eyeballs on the dependencies, so that is how I handle that now. That is one of the challenges, because it is not just parsing the source metadata, any existing spack package also has to be considered. Also, more specific to R is that any "system" dependencies are not defined in a consistent way in the metadata. The R build system makes no effort to parse that field when building. If the system dependency does not exist, the build will simply fail at build time. 

There are a couple of other points about R packages that I will mention. One thing that is common when updating R packages is that new r- packages (those not in spack) are needed. That tends to break the workflow as those packages have to be created. Having a better template would go a long way to help with that, but being able to auto generate them in the update scripts would be even better. A side effect of this tendency of R packages is that submitting R CRAN packages in the 1 package per PR model tends to leave quite a few things failing the PR audit check, which then has to be managed. I think SpackBot could help with that. Finally, while Bioconductor packages are each valid R packages, there is the extra level of the Bioconductor release version. That pretty much dictates that all Bioconductor packages are updated at roughly the same time and need to be updated in Spack as a single PR. Having new Spack Bioconductor packages deployed ahead of others can lead to a very broken Bioconductor stack at runtime for the end users, until everything catches up.

I usually work on the R package updates when I begin work on a new software stack release and I should be beginning that process fairly soon. I will work on those scripts a bit more at that time and see if they can be made more automated. I wonder if Jinja 2 could help with that. I need to look into that more because so far I have only used Jinja 2 to customize my module template.

**Reply by adamjstewart** [reply]
Yes, R will definitely have its differences with Python and other build systems. The idea here is that the build system-specific stuff (extracting the necessary metadata) can be separated from the Spack stuff (automatic package creation, checksumming, adding new versions/packages, etc.). I don't just want a script in the Spack repo that users can run, I want to integrate directly into `spack create` and `spack checksum`.

**Comment by cosmicexplorer** [comment]
@adamjstewart advised me over slack to link this here (https://github.com/pypa/pip/pull/10771): `pip install --report` now provides a json report with transitive dependency info! It will likely get faster when I follow up with the rest of https://github.com/pypa/pip/pull/10748, but this is the first time pip provides I believe exactly the information we would need to enable e.g. `spack install py-X` for a package without a specific recipe yet, or to update dependency info + more within `spack checksum`!

**Reply by adamjstewart** [reply]
Does this give us information about `setup_requires` vs. `install_requires`, or is it only the latter? I would have to experiment with this, but I'm guessing we would only need to look at the root spec and parse the list of `requires_dist`/`requires_python` to get what we need for Spack.

@pradyunsg may be interested in this discussion

**Reply by cosmicexplorer** [reply]
> Does this give us information about setup_requires vs. install_requires, or is it only the latter?

From about 12 minutes of searching, I am pretty sure that the metadata recorded (either from a wheel `METADATA` or a `PKG-INFO`) does *not* contain `setup_requires`. I found [PEP 518](https://peps.python.org/pep-0518/) which seems to describe that `setup_requires` is very difficult to use correctly anyway (it describes the current `pyproject.toml` standard for setup reqs), and [the latest setuptools docs](https://setuptools.pypa.io/en/latest/userguide/dependency_management.html) very clearly describe `setup_requires` as deprecated.

Since spack is already in the business of downloading and then immediately executing foreign code, I think that it might not be unreasonable to literally just execute the `setup.py` to get that info, though (although it is quite fun to think about how to symbolically execute `setup.py` files).

> I'm guessing we would only need to look at the root spec and parse the list of requires_dist/requires_python to get what we need for Spack.

I think you're probably right; I've been idly thinking about improving `spack checksum` for cargo packages the last few days and it was super neat to find we already have very well-developed thoughts on that front.

**Reply by adamjstewart** [reply]
`setup_requires` is deprecated in favor of `[build-system] requires` but I'm using them interchangeably here. Basically, I'm wondering if it's possible to find all `type='build'` deps programmatically. I guess we would either have to rely on wheels where no build deps are needed or ship our own toml parser (don't think we'll be dropping support for Python 3.10 and older any time soon) and also run `setup.py` for older packages.

**Reply by cosmicexplorer** [reply]
That is currently what I'm thinking, boooooo. At least it's tractable.

**Reply by pradyunsg** [reply]
> Does this give us information about `setup_requires` vs. `install_requires`, or is it only the latter?

Only the runtime dependencies.

The report is designed to show what would get *installed*, and it doesn't contain any information about build dependencies in the build environments for the source distributions for example.

**Comment by EbiArnie** [comment]
@adamjstewart recommended I put a link here to a script I created:
[Create Spack packages for Perl](https://github.com/EbiArnie/spack-contrib/blob/main/scripts/package-cpan-api-spack.pl).
```console
$ package-cpan-api-spack.pl Test::NoWarnings
```
will create ready-to-use Spack package descriptions for you.

**Reply by EbiArnie** [reply]
In a previous iteration, this script would only provide the dependencies and basically call 'spack create https://...'
But that was slow due to the manual steps. Also, since we already have everything in the metadata (including the checksum), it's easy to create the full package.py.
This has the advantage that the translation from Perl to Spack names (e.g. LWP::Simple to perl-libwww-perl) is made by the script and I don't have to remember those. And I can directly generate all dependencies.
